core/src/main/scala/org/apache/spark/sql/extensions/rules.scala

```diff
@@ -55,22 +56,45 @@ case class TiResolutionRule(getOrCreateTiContext: SparkSession => TiContext)(
         tiSession,
         TiTableReference(dbName, tableName, sizeInBytes),
         meta,
-        Some(ts)
+        Option.apply(ts)
       )(sqlContext)
       // Use SubqueryAlias so that projects and joins can correctly resolve
       // UnresolvedAttributes in JoinConditions, Projects, Filters, etc.
       SubqueryAlias(tableName, LogicalRelation(tiDBRelation))
     }
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
-    val ts = tiContext.tiSession.getTimestamp
-    plan transformUp {
+    val ts = tiSession.getTimestamp
+
+    def applyTimestamp: PartialFunction[LogicalPlan, LogicalPlan] = {
       case UnresolvedRelation(tableIdentifier)
           if tiCatalog
             .catalogOf(tableIdentifier.database)
             .exists(_.isInstanceOf[TiSessionCatalog]) =>
         resolveTiDBRelation(tableIdentifier, ts)
+      case f @ Filter(condition, _) =>
+        f.copy(
+          condition = condition.transform {
+            case e @ Exists(p, _, _) =>
+              e.copy(plan = p.transform(applyTimestamp))
+            case ls @ ListQuery(p, _, _, _) =>
+              ls.copy(plan = p.transform(applyTimestamp))
+            case s @ ScalarSubquery(p, _, _) =>
+              s.copy(plan = p.transform(applyTimestamp))
+          }
+        )
+      case p @ Project(projectList, _) =>
```

Comment:
is there any other cases (except `UnresolvedRelation`, `Filter` and `Project`)?

Reply:
It uses same logic as TiResolutionRule. If it breaks the resolution rule will break first.

Reply:
It uses same logic as TiResolutionRule. If it breaks the resolution rule will break first.